### PR TITLE
Wait with push to master until PR is updated

### LINKED ISSUE #196: PRs are not always correctly closed as merged
Occasionally, it happens that a PR merged by hoff ends up showing as closed rather than merged, i.e.
image
rather than
image

And sometimes, github just doesn't close the PR at all.

In hoff, the merging procedure is defined here:

https://github.com/channable/hoff/blob/46805f05a0c23ee1bac5be786da1acd23d076ce3/src/Logic.hs#L203-L210

We first force-push the rebased version of the original branch, and then push the merge to master.

That way, GitHub *should* see that the commits in the PR branch are exactly those that are on master, and correctly close the PR as merged. This is just like a human would do a rebase & merge via the `git` command line (i.e. not using GitHub's interface for that), albeit a lot faster.

The latter point might actually be the issue here. My hypothesis is that if the force push of the PR branch and the push to master are sent in rapid succession, GitHub might become confused about the state of the branches, and incorrectly observe a state where the commits that were merged are not yet on the PR branch.

A hacky band-aid could be to just sleep for a few seconds between force-pushing the PR branch and pushing master.

### PR DESCRIPTION
**Pull request by musicismyalibi**
After force pushing to the PR branch we wait until we receive the GitHub event that the commit for the PR has changed before pushing to master. If this event takes too long to arrive we push to master anyway.

Resolves #196

### CODE REVIEW

**Comment by ruuda** [comment]
I take it #245 did not solve the problem then?

File: src/Logic.hs [hunk 1]
```diff
@@ -1186,51 +1246,36 @@ pushCandidate (pullRequestId, pullRequest) newHead@(Sha sha) state =
...
+              mChangelog <- getChangelog previousTag newHead
+              let
+                tagName = versionToTag $ v + 1
+                changelog = fromMaybe "Failed to get the changelog" mChangelog
+                tagMessage = messageForTag tagName approvalKind changelog
+              pushResult <- tryForcePush prBranch newHead
+              case pushResult of
+                PushOk -> pure $ Pr.setIntegrationStatus pullRequestId (Pr.PromoteAndTag currTime newHead tagName tagMessage) state
+                PushRejected _ -> tryIntegratePullRequest pullRequestId
+                                $ unintegrateAfter pullRequestId state
```
Comment (crtschin) [inline src/Logic.hs:1272]: It looks like this is duplicated 3 times in this function.

File: src/Logic.hs [hunk 2]
```diff
@@ -1186,51 +1246,36 @@ pushCandidate (pullRequestId, pullRequest) newHead@(Sha sha) state =
...
-                    then "It is scheduled for autodeploy!"
-                    else Text.concat ["Please wait for the build of ", sha, " to pass and don't forget to deploy it!"]
-            pure pushResult
-      else
-        tryPromote prBranch newHead
-    case pushResult of
-      -- If the push worked, then this was the final stage of the pull request.
-      -- GitHub will mark the pull request as closed, and when we receive that
-      -- event, we delete the pull request from the state. Until then, reset
-      -- the integration candidate, so we proceed with the next pull request.
```
Comment (crtschin) [inline src/Logic.hs:1221]: You removed quite some comments. If they're outdated relative to the implementation, can you update them instead?

File: tests/Spec.hs [hunk 3]
```diff
@@ -1877,10 +1894,55 @@ main = hspec $ do
...
+       in actions `shouldBe`
+            [ AIsReviewer (Username "deckard")
+            , ALeaveComment prId "<!-- Hoff: ignore -->\nPull request approved for merge by @deckard, rebasing now."
+            , ATryIntegrate "Merge #1: Untitled\n\nApproved-by: deckard\nAuto-deploy: false\n" (PullRequestId 1, Branch "refs/pull/1/head", Sha "abc1234") [] False
+            , ALeaveComment prId "<!-- Hoff: ignore -->\nRebased as def2345, waiting for CI \8230"
+            , ALeaveComment prId "<!-- Hoff: ignore -->\n[CI job :yellow_circle:](example.com/1b2) started."
+            , ATryForcePush (Branch "p") (Sha "def2345")
+            ]
+
+    it "promotes if timeout has expired" $
```
Comment (crtschin) [inline tests/Spec.hs:1923]: Interestingly, by detaching the rebase (force-)push from the push to the 'main'
branch into separate steps. It is now possible that hoff handles, and perhaps
rebases any other PR that may have been `@bot merge`d. If i'm not mistaken the
current implementation will simply cause one of the two merge's to fail. How
this interacts with the merge train functionality is also quite confusing.

Can you add a testcase for this, and see if this is a problem?

**Review by crtschin — CHANGES_REQUESTED** [review]
I'm not super sure about this, I added my concerns

File: tests/Spec.hs (hunk 3 above)
Comment (crtschin) [reply tests/Spec.hs:1923]: Was explained offline. What happens is that the subsequent `@bot merge` wil put the PR in the merge train and will wait until the previous merge gets pushed to master.

File: src/Logic.hs (hunk 2 above)
Comment (musicismyalibi) [reply src/Logic.hs:1221]: These comments are not actually removed, but only moved, together with the code, to the promoting part of the process.

File: tests/Spec.hs [hunk 4]
```diff
@@ -1877,7 +1894,278 @@ main = hspec $ do
...
+        , ALeaveComment (PullRequestId 2) "<!-- Hoff: ignore -->\n[CI job :yellow_circle:](example.com/2b2) started."
+        , ATryForcePush (Branch "fst") (Sha "1b2")
+        , ATryPromote (Sha "1b2")
+        , ACleanupTestBranch (PullRequestId 1)
+        , ATryForcePush (Branch "snd") (Sha "2b2")
+        , ATryPromote (Sha "2b2")
+        , ACleanupTestBranch (PullRequestId 2)
+        ]
+
+    it "correctly a PR that is closed while waiting for promotion" $ do
```
Comment (crtschin) [inline tests/Spec.hs:2140]: I think you're missing a verb here

File: tests/Spec.hs [hunk 5]
```diff
@@ -1877,7 +1894,278 @@ main = hspec $ do
...
+          $ Project.emptyProjectState
+        events =
+          [ CommentAdded (PullRequestId 1) "deckard" "@bot merge"
+          , BuildStatusChanged (Sha "1b2") "default" (Project.BuildStarted "example.com/1b2")
+          , CommentAdded (PullRequestId 2) "deckard" "@bot merge"
+          , BuildStatusChanged (Sha "2b2") "default" (Project.BuildStarted "example.com/2b2")
+          , BuildStatusChanged (Sha "2b2") "default" Project.BuildSucceeded
+          , BuildStatusChanged (Sha "1b2") "default" Project.BuildSucceeded
+          , PullRequestCommitChanged (PullRequestId 1) (Sha "1b2")
+          , PullRequestCommitChanged (PullRequestId 2) (Sha "2b2")
```
Comment (crtschin) [inline tests/Spec.hs:2104]: So the pull request commit changed event does the push. What happens if you
reverse the order of these two? Does `2b2` (which contains the changes in `1b2`) get
merged into master first before `1b2`?

Comment (crtschin) [reply tests/Spec.hs:2106]: So only a single PR will ever be 'awaiting promotion'?

Comment (musicismyalibi) [reply tests/Spec.hs:2106]: Only a single PR per project, yes.

**Review by crtschin — APPROVED** [review]
I think this works

**Comment by musicismyalibi** [comment]
@OpsBotPrime merge

**Comment by OpsBotPrime** [comment]
Pull request approved for merge by @musicismyalibi, rebasing now.

**Comment by OpsBotPrime** [comment]
Rebased as e0c169e7992c3304912a85ea341e08cf184356b8, waiting for CI …

**Comment by OpsBotPrime** [comment]
[CI job :yellow_circle:](https://channable.semaphoreci.com/workflows/4680629a-c595-498a-b372-1807678b0a42?pipeline_id=3de9300b-7712-431d-833f-60f73f10740c) started.

**Pull request merged**